tools/asset-sync/assets.ps1

```diff
@@ -0,0 +1,521 @@
+
+.PARAMETER Config
+A PSCustomObject that contains an auto-parsed assets.json content from Resolve-AssetsJson.
+
+#>
+Function Push-AssetsRepo-Update {
+    param(
+        [PSCustomObject]$Config
+    )
+    $newSha = $Config.SHA
+    $gitUser = git config --global user.name
+    $autoCommitMessage = "Automatic asset update from $gitUser."
+
+    $assetRepo = Resolve-AssetRepo-Location -Config $Config
+    try {
+        Push-Location $assetRepo
+        $statusResult = git status --porcelain
+
+        if(!$statusResult){
+            Write-Host "No pending changes."
+            exit(0)
+        }
+
+        $alreadyLatestSHA = $true
+        Write-Host "git rev-parse origin/$($Config.AssetsRepoBranch)"
+        $retrievedLatestSHA = git rev-parse origin/$($Config.AssetsRepoBranch)
+        Write-Host "Latest SHA is $retrievedLatestSHA."
+
+        # if the above command fails with code 128, the target auto commit branch does not exist, and we need to create it
+        if($LASTEXITCODE -ne 0 -and $LASTEXITCODE -eq 128){
+            Write-Host "Need to checkout new branch based on current changes."
+            $alreadyLatestSHA = $true
+        }
+        elseif ($LASTEXITCODE -ne 0) {
+            Write-Error "A non-code-128 error is not expected here. Check git command output above."
+            exit(1)
+        }
+        # if the branch already exists, we need to check to see if we're actually on the latest commit 
+        else {
+            if($retrievedLatestSHA -ne $Config.SHA){
+                $alreadyLatestSHA = $false
+            }
+        }
+
+        Write-Host "Based off latest commit: $alreadyLatestSHA"
+        
+        # if we are based off the latest commit (or it's a nonexistent branch), all we gotta do is checkout a new branch of the correct name and push it.
+        if($alreadyLatestSHA) {
+            Write-Host "git checkout -b $($Config.AssetsRepoBranch)"
+            git checkout -b $($Config.AssetsRepoBranch)
+            Write-Host "git add -A ."
+            git add -A .
+            Write-Host "git commit -m `"$autoCommitMessage`""
+            git commit -m "$autoCommitMessage"
+            Write-Host "git push origin $($Config.AssetsRepoBranch)"
+            git push origin $($Config.AssetsRepoBranch)
+        }
+        else {
+            # TODO is there a noticable downside to stash versus saving our own patchfile like in git-branch-push?
+            Write-Host "git stash"
+            git stash
+
+            # TODO we want to only fetch the latest commit, instead of the entire branch
+            Write-Host "git fetch origin $($Config.AssetsRepoBranch)"
+            git fetch origin $($Config.AssetsRepoBranch)
+            Write-Host "git checkout $($Config.AssetsRepoBranch)"
+            git checkout $($Config.AssetsRepoBranch)
+
+            Write-Host "git stash pop"
+            git stash pop
```

Comment:
Are there cases where this may run into merge/stash conflicts?

Reply:
And do we need any handling to avoid cases where we commit and push stash conflicts from git like:

```
<<<<<<< Updated upstream
foobarupstream
=======
foobarpoppedstash
>>>>>>> Stashed changes
```

If the recordings are supposed to be json, maybe we prevent commits of files that are invalid json? Or have some set of validation middleware for the "add and commit" step.

Reply:
I was relying on git's conflict management to stop me. I need to add explicit error code handlings here.

I need to create a conflict scenario in the integration tests and see what that will look like.